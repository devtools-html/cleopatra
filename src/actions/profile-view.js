/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type {
  ProfileSelection, CallTreeFilter, ImplementationFilter,
} from '../types/actions';
import type { Action, ThunkAction } from '../types/store';
import type { ThreadIndex, IndexIntoFuncTable, IndexIntoMarkersTable } from '../types/profile';

/**
 * The actions that pertain to changing the view on the profile, including searching
 * and filtering. Currently the call tree's actions are in this file, but should be
 * split apart. These actions should most likely affect every panel.
 */
export function changeSelectedFuncStack(
  threadIndex: ThreadIndex,
  selectedFuncStack: IndexIntoFuncTable[]
): Action {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack, threadIndex,
  };
}

export function changeSelectedThread(selectedThread: ThreadIndex): Action {
  return {
    type: 'CHANGE_SELECTED_THREAD',
    selectedThread,
  };
}

export function changeThreadOrder(threadOrder: ThreadIndex[]): Action {
  return {
    type: 'CHANGE_THREAD_ORDER',
    threadOrder,
  };
}

export function hideThread(
  threadIndex: ThreadIndex,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[]
): ThunkAction<void> {
  return dispatch => {
    // Do not allow hiding the last thread.
    if (hiddenThreads.length + 1 === threadOrder.length) {
      return;
    }

    dispatch(({
      type: 'HIDE_THREAD',
      threadIndex,
      threadOrder,
      hiddenThreads,
    }: Action));
  };
}

export function showThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'SHOW_THREAD',
    threadIndex,
  };
}

export function changeCallTreeSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_CALL_TREE_SEARCH_STRING',
    searchString,
  };
}

export function changeExpandedFuncStacks(
  threadIndex: ThreadIndex,
  expandedFuncStacks: Array<IndexIntoFuncTable[]>
): Action {
  return {
    type: 'CHANGE_EXPANDED_FUNC_STACKS',
    threadIndex, expandedFuncStacks,
  };
}

export function changeSelectedMarker(
  threadIndex: ThreadIndex,
  selectedMarker: IndexIntoMarkersTable | -1
): Action {
  return {
    type: 'CHANGE_SELECTED_MARKER',
    selectedMarker, threadIndex,
  };
}

export function changeImplementationFilter(implementation: ImplementationFilter): Action {
  return {
    type: 'CHANGE_IMPLEMENTATION_FILTER',
    implementation,
  };
}

export function changeInvertCallstack(invertCallstack: boolean): Action {
  return {
    type: 'CHANGE_INVERT_CALLSTACK',
    invertCallstack,
  };
}

export function changeHidePlatformDetails(hidePlatformDetails: boolean): Action {
  return {
    type: 'CHANGE_HIDE_PLATFORM_DETAILS',
    hidePlatformDetails,
  };
}

export type UpdateProfileSelection = (selection: ProfileSelection) => Action;

export function updateProfileSelection(selection: ProfileSelection): Action {
  return {
    type: 'UPDATE_PROFILE_SELECTION',
    selection,
  };
}

export function addRangeFilter(start: number, end: number): Action {
  return {
    type: 'ADD_RANGE_FILTER',
    start, end,
  };
}

export function addRangeFilterAndUnsetSelection(start: number, end: number): ThunkAction<void> {
  return dispatch => {
    dispatch(addRangeFilter(start, end));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function popRangeFilters(firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_RANGE_FILTERS',
    firstRemovedFilterIndex,
  };
}

export function popRangeFiltersAndUnsetSelection(firstRemovedFilterIndex: number): ThunkAction<void> {
  return dispatch => {
    dispatch(popRangeFilters(firstRemovedFilterIndex));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function addCallTreeFilter(threadIndex: ThreadIndex, filter: CallTreeFilter): Action {
  return {
    type: 'ADD_CALL_TREE_FILTER',
    threadIndex,
    filter,
  };
}

export function popCallTreeFilters(threadIndex: ThreadIndex, firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_CALL_TREE_FILTERS',
    threadIndex,
    firstRemovedFilterIndex,
  };
}
