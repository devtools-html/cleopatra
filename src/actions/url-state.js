/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action } from '../types/store';

export function urlStateHasChanged(): Action {
  return {
    type: 'URL_STATE_HAS_CHANGED',
  };
}
