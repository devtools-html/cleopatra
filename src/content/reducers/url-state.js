// @flow
import { combineReducers } from 'redux';
import { defaultThreadOrder } from '../profile-data';
import { createSelector } from 'reselect';
import { urlFromState } from '../url-handling';
import * as RangeFilters from '../range-filters';

import type { ThreadIndex } from '../../common/types/profile';
import type { StartEndRange } from '../../common/types/units';
import type { Action, CallTreeFiltersPerThread, CallTreeFilter, DataSource } from '../actions/types';
import type { State, URLState, Reducer } from './types';

function dataSource(state: DataSource = 'none', action: Action) {
  switch (action.type) {
    case 'WAITING_FOR_PROFILE_FROM_FILE':
      return 'from-file';
    case 'PROFILE_PUBLISHED':
      return 'public';
    default:
      return state;
  }
}

function hash(state: string = '', action: Action) {
  switch (action.type) {
    case 'PROFILE_PUBLISHED':
      return action.hash;
    default:
      return state;
  }
}

function selectedTab(state: string = 'calltree', action: Action) {
  switch (action.type) {
    case 'CHANGE_SELECTED_TAB':
      return action.selectedTab;
    default:
      return state;
  }
}

function rangeFilters(state: StartEndRange[] = [], action: Action) {
  switch (action.type) {
    case 'ADD_RANGE_FILTER': {
      const { start, end } = action;
      return [...state, { start, end }];
    }
    case 'POP_RANGE_FILTERS':
      return state.slice(0, action.firstRemovedFilterIndex);
    default:
      return state;
  }
}

function selectedThread(state: ThreadIndex = 0, action: Action) {
  switch (action.type) {
    case 'CHANGE_SELECTED_THREAD':
      return action.selectedThread;
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_FILE': {
      // When loading in a brand new profile, select either the GeckoMain [tab] thread,
      // or the first thread in the thread order. For profiles from the Web, the
      // selectedThread has already been initialized from the URL and does not require
      // looking at the profile.
      const contentThreadId = action.profile.threads.findIndex(thread => {
        return thread.name === 'GeckoMain' && thread.processType === 'tab';
      });
      return contentThreadId !== -1 ? contentThreadId : defaultThreadOrder(action.profile.threads)[0];
    }
    default:
      return state;
  }
}

function callTreeSearchString(state: string = '', action: Action) {
  switch (action.type) {
    case 'CHANGE_CALL_TREE_SEARCH_STRING':
      return action.searchString;
    default:
      return state;
  }
}

function callTreeFilters(state: CallTreeFiltersPerThread = {}, action: Action) {
  switch (action.type) {
    case 'ADD_CALL_TREE_FILTER': {
      const { threadIndex, filter } = action;
      const oldFilters = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: [...oldFilters, filter],
      });
    }
    case 'POP_CALL_TREE_FILTERS': {
      const { threadIndex, firstRemovedFilterIndex } = action;
      const oldFilters = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: oldFilters.slice(0, firstRemovedFilterIndex),
      });
    }
    default:
      return state;
  }
}

function jsOnly(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_JS_ONLY':
      return action.jsOnly;
    default:
      return state;
  }
}

function invertCallstack(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_INVERT_CALLSTACK':
      return action.invertCallstack;
    default:
      return state;
  }
}

function hidePlatformDetails(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_HIDE_PLATFORM_DETAILS':
      return action.hidePlatformDetails;
    default:
      return state;
  }
}

const urlStateReducer: Reducer<URLState> = (regularUrlStateReducer => (state: URLState, action: Action): URLState => {
  switch (action.type) {
    case '@@urlenhancer/updateURLState':
      return action.urlState;
    default:
      return regularUrlStateReducer(state, action);
  }
})(combineReducers({
  dataSource, hash, selectedTab, rangeFilters, selectedThread,
  callTreeSearchString, callTreeFilters, jsOnly, invertCallstack,
  hidePlatformDetails,
}));
export default urlStateReducer;

const getURLState = (state: State): URLState => state.urlState;

export const getDataSource = (state: State) => getURLState(state).dataSource;
export const getHash = (state: State) => getURLState(state).hash;
export const getRangeFilters = (state: State) => getURLState(state).rangeFilters;
export const getJSOnly = (state: State) => getURLState(state).jsOnly;
export const getHidePlatformDetails = (state: State) => getURLState(state).hidePlatformDetails;
export const getInvertCallstack = (state: State) => getURLState(state).invertCallstack;
export const getSearchString = (state: State) => getURLState(state).callTreeSearchString;
export const getSelectedTab = (state: State) => getURLState(state).selectedTab;
export const getSelectedThreadIndex = (state: State) => getURLState(state).selectedThread;
export const getCallTreeFilters = (state: State, threadIndex: ThreadIndex): CallTreeFilter[] => {
  return getURLState(state).callTreeFilters[threadIndex] || [];
};

export const getURLPredictor = createSelector(
  getURLState,
  (oldURLState: URLState) => actionOrActionList => {
    const actionList = ('type' in actionOrActionList) ? [actionOrActionList] : actionOrActionList;
    const newURLState = actionList.reduce(urlStateReducer, oldURLState);
    return urlFromState(newURLState);
  }
);

export const getRangeFilterLabels = createSelector(
  getRangeFilters,
  RangeFilters.getRangeFilterLabels
);
