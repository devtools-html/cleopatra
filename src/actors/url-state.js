/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import { getUrlState } from '../reducers/url-state';
import {
  urlStateHasChanged,
  updateUrlState,
  show404,
} from '../actions/url-state';
import { urlFromState, stateFromLocation } from '../url-handling';

import type { Dispatch } from '../types/store';
import type { State, UrlState } from '../types/reducers';

let previousUrlState = null;
export default function stateWatcher(state: State, dispatch: Dispatch) {
  function updateState() {
    if (window.history.state) {
      dispatch(updateUrlState(window.history.state));
    } else {
      try {
        const urlState = stateFromLocation(window.location);
        // replacing the existing location ensures we get the latest format for the URL
        updateLocation(urlState, true);
        dispatch(updateUrlState(urlState));
      } catch (e) {
        console.error(e);
        show404(window.location.pathname + window.location.search);
      }
    }
  }

  function initialSetup() {
    updateState();
    window.addEventListener('popstate', updateState);
  }

  function updateLocation(urlState: UrlState, replace: boolean = false) {
    const newUrl = urlFromState(urlState);
    const operation = replace
      ? window.history.replaceState.bind(window.history)
      : window.history.pushState.bind(window.history);
    if (newUrl !== window.location.pathname + window.location.search) {
      operation(urlState, document.title, newUrl);
    }
  }

  const newUrlState = getUrlState(state);
  if (!previousUrlState) {
    // first run
    previousUrlState = newUrlState;
    initialSetup();
  } else if (newUrlState !== previousUrlState) {
    previousUrlState = newUrlState;
    updateLocation(newUrlState);
    dispatch(urlStateHasChanged());
  }
}
