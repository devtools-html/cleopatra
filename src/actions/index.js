/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';
import * as profileUpload from './profile-upload';
import * as shortUrl from './short-url';
import * as urlState from './url-state';

export default Object.assign(
  {},
  app,
  icons,
  profileSummary,
  profileUpload,
  profileView,
  receiveProfile,
  shortUrl,
  timeline,
  urlState
);
