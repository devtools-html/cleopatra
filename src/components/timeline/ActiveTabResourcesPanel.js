/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import classNames from 'classnames';
import explicitConnect from '../../utils/connect';
import { withSize } from '../shared/WithSize';
import { getIsActiveTabResourcesPanelOpen } from '../../selectors/url-state';
import { toggleResourcesPanel } from '../../actions/app';
import { ACTIVE_TAB_TIMELINE_RESOURCES_HEADER_HEIGHT } from '../../app-logic/constants';

import type { SizeProps } from '../shared/WithSize';
import type { LocalTrack } from '../../types/profile-derived';
import type { ConnectedProps } from '../../utils/connect';

type OwnProps = {|
  +resourceTracks: LocalTrack[],
  +setIsInitialSelectedPane: (value: boolean) => void,
|};

type StateProps = {|
  isActiveTabResourcesPanelOpen: boolean,
|};

type DispatchProps = {|
  +toggleResourcesPanel: typeof toggleResourcesPanel,
|};

type Props = {|
  ...SizeProps,
  ...ConnectedProps<OwnProps, StateProps, DispatchProps>,
|};

class ActiveTabResourcesPanel extends React.PureComponent<Props> {
  render() {
    const {
      resourceTracks,
      toggleResourcesPanel,
      isActiveTabResourcesPanelOpen,
    } = this.props;
    return (
      <div
        className="timelineResources"
        style={{
          '--resources-header-height': `${ACTIVE_TAB_TIMELINE_RESOURCES_HEADER_HEIGHT}px`,
        }}
      >
        <div
          onClick={toggleResourcesPanel}
          className={classNames('timelineResourcesHeader', {
            opened: isActiveTabResourcesPanelOpen,
          })}
        >
          Resources ({resourceTracks.length})
        </div>
        {isActiveTabResourcesPanelOpen ? (
          <ol className="timelineResourceTracks">
            {/* TODO: Add the Resource tracks here */}
          </ol>
        ) : null}
      </div>
    );
  }
}

export default explicitConnect<OwnProps, StateProps, DispatchProps>({
  mapStateToProps: state => ({
    isActiveTabResourcesPanelOpen: getIsActiveTabResourcesPanelOpen(state),
  }),
  mapDispatchToProps: { toggleResourcesPanel },
  component: withSize<Props>(ActiveTabResourcesPanel),
});
