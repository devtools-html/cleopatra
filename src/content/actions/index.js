// @flow
import * as app from './app';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';

export default Object.assign({}, app, profileSummary, profileView, receiveProfile, timeline);
