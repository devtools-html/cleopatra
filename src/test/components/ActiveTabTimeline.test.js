/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import ReactDOM from 'react-dom';
import Timeline from '../../components/timeline';
import ActiveTabGlobalTrack from '../../components/timeline/ActiveTabGlobalTrack';
import ActiveTabResourcesPanel from '../../components/timeline/ActiveTabResourcesPanel';
import { render, fireEvent } from 'react-testing-library';
import { Provider } from 'react-redux';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileWithNiceTracks } from '../fixtures/profiles/tracks';
import { changeTimelineTrackOrganization } from '../../actions/receive-profile';
import { getBoundingBox } from '../fixtures/utils';
import { addActiveTabInformationToProfile } from '../fixtures/profiles/processed-profile';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import {
  getActiveTabGlobalTracks,
  getActiveTabResourceTracks,
} from '../../selectors/profile';
import { getSelectedThreadIndex } from '../../selectors/url-state';
import { changeSelectedThread } from '../../actions/profile-view';
import { ensureExists } from '../../utils/flow';

describe('ActiveTabTimeline', function() {
  beforeEach(() => {
    jest.spyOn(ReactDOM, 'findDOMNode').mockImplementation(() => {
      // findDOMNode uses nominal typing instead of structural (null | Element | Text), so
      // opt out of the type checker for this mock by returning `any`.
      const mockEl = ({
        getBoundingClientRect: () => getBoundingBox(300, 300),
      }: any);
      return mockEl;
    });

    jest
      .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
      .mockImplementation(() => getBoundingBox(200, 300));

    const ctx = mockCanvasContext();
    jest
      .spyOn(HTMLCanvasElement.prototype, 'getContext')
      .mockImplementation(() => ctx);
  });

  it('should be rendered properly from the Timeline component', () => {
    const {
      profile,
      parentInnerWindowIDsWithChildren,
      firstTabBrowsingContextID,
    } = addActiveTabInformationToProfile(getProfileWithNiceTracks());
    profile.threads[0].frameTable.innerWindowID[0] = parentInnerWindowIDsWithChildren;
    const store = storeWithProfile(profile);
    store.dispatch(
      changeTimelineTrackOrganization({
        type: 'active-tab',
        browsingContextID: firstTabBrowsingContextID,
      })
    );

    const { container } = render(
      <Provider store={store}>
        <Timeline />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('ActiveTabGlobalTrack', function() {
    function setup() {
      const { profile, ...pageInfo } = addActiveTabInformationToProfile(
        getProfileWithNiceTracks()
      );
      profile.threads[0].frameTable.innerWindowID[0] =
        pageInfo.parentInnerWindowIDsWithChildren;
      const store = storeWithProfile(profile);
      store.dispatch(
        changeTimelineTrackOrganization({
          type: 'active-tab',
          browsingContextID: pageInfo.firstTabBrowsingContextID,
        })
      );
      const trackIndex = 0;
      const { getState, dispatch } = store;
      const trackReference = { type: 'global', trackIndex };
      const tracks = getActiveTabGlobalTracks(getState());
      const track = tracks[trackIndex];
      const setInitialSelected = () => {};
      if (track.type !== 'tab') {
        throw new Error('Expected a tab track.');
      }
      const threadIndex = track.threadIndex;

      if (threadIndex !== null) {
        // The assertions are simpler if the GeckoMain tab thread is not already selected.
        dispatch(changeSelectedThread(threadIndex + 1));
      }

      const renderResult = render(
        <Provider store={store}>
          <ActiveTabGlobalTrack
            trackIndex={trackIndex}
            trackReference={trackReference}
            setInitialSelected={setInitialSelected}
          />
        </Provider>
      );
      const { container } = renderResult;

      const getGlobalTrackRow = () =>
        ensureExists(
          container.querySelector('.timelineTrackGlobalRow'),
          `Couldn't find the track global row with selector .timelineTrackGlobalRow`
        );

      return {
        ...renderResult,
        ...pageInfo,
        dispatch,
        getState,
        profile,
        store,
        trackReference,
        trackIndex,
        threadIndex,
        getGlobalTrackRow,
      };
    }

    it('matches the snapshot of a global tab track', () => {
      const { container } = setup();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('has useful parts of the component', function() {
      const { getGlobalTrackRow } = setup();
      expect(getGlobalTrackRow()).toBeTruthy();
    });

    it('starts out not being selected', function() {
      const { getState, getGlobalTrackRow, threadIndex } = setup();
      expect(getSelectedThreadIndex(getState())).not.toBe(threadIndex);
      expect(getGlobalTrackRow().classList.contains('selected')).toBe(false);
    });

    it('can select a thread by clicking the row', () => {
      const { getState, getGlobalTrackRow, threadIndex } = setup();
      expect(getSelectedThreadIndex(getState())).not.toBe(threadIndex);
      fireEvent.click(getGlobalTrackRow());
      expect(getSelectedThreadIndex(getState())).toBe(threadIndex);
    });

    it('does not display the resources panel if there are no resource tracks', () => {
      const { getState, queryByText } = setup();
      expect(getActiveTabResourceTracks(getState()).length).toBe(0);
      expect(queryByText(/Resources/)).toBe(null);
    });
  });

  describe('ActiveTabResourcesPanel', function() {
    function setup() {
      const { profile, ...pageInfo } = addActiveTabInformationToProfile(
        getProfileWithNiceTracks()
      );
      // Setting the first thread as parent track and the second as the iframe track.
      profile.threads[0].frameTable.innerWindowID[0] =
        pageInfo.parentInnerWindowIDsWithChildren;
      profile.threads[1].frameTable.innerWindowID[0] =
        pageInfo.iframeInnerWindowIDsWithChild;
      const store = storeWithProfile(profile);
      store.dispatch(
        changeTimelineTrackOrganization({
          type: 'active-tab',
          browsingContextID: pageInfo.firstTabBrowsingContextID,
        })
      );
      const { getState, dispatch } = store;
      const resourceTracks = getActiveTabResourceTracks(getState());

      const renderResult = render(
        <Provider store={store}>
          <ActiveTabResourcesPanel
            resourceTracks={resourceTracks}
            setIsInitialSelectedPane={() => {}}
          />
        </Provider>
      );

      const { getByText } = renderResult;
      const getResourcesPanelHeader = () => getByText(/Resources/);

      return {
        ...renderResult,
        ...pageInfo,
        dispatch,
        getState,
        profile,
        store,
        getResourcesPanelHeader,
      };
    }

    it('matches the snapshot of a resources panel', () => {
      const { container } = setup();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('is closed by default', () => {
      const { getResourcesPanelHeader } = setup();
      // TODO: Currently it's not possible to test this without accessing the class
      // directly but this is not ideal for testing. We should assert user-like actions
      // instead when we have content in this panel.
      expect(getResourcesPanelHeader().classList.contains('opened')).toBe(
        false
      );
    });

    it('clicking on the header opens the resources panel', () => {
      const { getResourcesPanelHeader } = setup();
      const resourcesPanelHeader = getResourcesPanelHeader();
      // TODO: Currently it's not possible to test this without accessing the class
      // directly but this is not ideal for testing. We should assert user-like actions
      // instead when we have content in this panel.
      expect(resourcesPanelHeader.classList.contains('opened')).toBe(false);

      fireEvent.click(resourcesPanelHeader);
      expect(resourcesPanelHeader.classList.contains('opened')).toBe(true);
    });
  });
});
