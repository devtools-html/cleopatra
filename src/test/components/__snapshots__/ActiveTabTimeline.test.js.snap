// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveTabTimeline ActiveTabGlobalTrack matches the snapshot of a global tab track 1`] = `
<li
  class="timelineTrack"
>
  <div
    class="timelineTrackRow timelineTrackGlobalRow activeTab"
  >
    <div
      class="timelineTrackTrack"
    >
      <div
        class="timelineTrackThread"
      >
        <div
          class="timelineMarkers"
          data-testid="TimelineMarkersJank"
        >
          <div
            class="react-contextmenu-wrapper"
          >
            <canvas
              class="timelineMarkersCanvas"
            />
          </div>
        </div>
        <div
          class="timelineMarkers timelineMarkersGeckoMain"
          data-testid="TimelineMarkersOverview"
        >
          <div
            class="react-contextmenu-wrapper"
          >
            <canvas
              class="timelineMarkersCanvas"
            />
          </div>
        </div>
        <div
          class="threadActivityGraph"
        >
          <canvas
            class="threadActivityGraphCanvas threadActivityGraphCanvas"
            height="300"
            width="200"
          />
        </div>
        <div
          class="timelineEmptyThreadIndicator"
        />
      </div>
    </div>
  </div>
</li>
`;

exports[`ActiveTabTimeline ActiveTabResourcesPanel matches the snapshot of a resources panel 1`] = `
<div
  class="timelineResources"
  style="--resources-header-height: 20px;"
>
  <div
    class="timelineResourcesHeader"
  >
    Resources (
    1
    )
  </div>
</div>
`;

exports[`ActiveTabTimeline should be rendered properly from the Timeline component 1`] = `
<div
  class="timelineSelection activeTab"
>
  <div
    class="timelineRuler"
    style="--timeline-ruler-height: 20px;"
  >
    <ol
      class="timelineRulerContainer"
    />
  </div>
  <div
    class="overflowEdgeIndicator timelineOverflowEdgeIndicator"
  >
    <div
      class="overflowEdgeIndicatorEdge topEdge"
    />
    <div
      class="overflowEdgeIndicatorEdge rightEdge"
    />
    <div
      class="overflowEdgeIndicatorEdge bottomEdge"
    />
    <div
      class="overflowEdgeIndicatorEdge leftEdge"
    />
    <div
      class="overflowEdgeIndicatorScrollbox timelineOverflowEdgeIndicatorScrollbox"
    >
      <div
        class="overflowEdgeIndicatorContentsWrapper"
      >
        <ol
          class="timelineThreadList"
        >
          <li
            class="timelineTrack"
          >
            <div
              class="timelineTrackRow timelineTrackGlobalRow activeTab selected"
            >
              <div
                class="timelineTrackTrack"
              >
                <div
                  class="timelineTrackThread"
                >
                  <div
                    class="timelineMarkers selected"
                    data-testid="TimelineMarkersJank"
                  >
                    <div
                      class="react-contextmenu-wrapper"
                    >
                      <canvas
                        class="timelineMarkersCanvas"
                      />
                    </div>
                  </div>
                  <div
                    class="timelineMarkers timelineMarkersGeckoMain selected"
                    data-testid="TimelineMarkersOverview"
                  >
                    <div
                      class="react-contextmenu-wrapper"
                    >
                      <canvas
                        class="timelineMarkersCanvas"
                      />
                    </div>
                  </div>
                  <div
                    class="threadActivityGraph"
                  >
                    <canvas
                      class="threadActivityGraphCanvas threadActivityGraphCanvas"
                      height="300"
                      width="200"
                    />
                  </div>
                  <div
                    class="timelineEmptyThreadIndicator"
                  />
                </div>
              </div>
            </div>
          </li>
        </ol>
      </div>
    </div>
  </div>
  <div
    class="timelineSelectionHoverLine"
    style="visibility: hidden; left: 0px;"
  />
</div>
`;
